### PR TITLE
Make compare previous better + new legend

### PR DESCRIPTION
**Pull request by alexkim205**
## Changes

The new legend is behind a FF called `insight-legends`

- [x] Update insight canvas header styles to match https://github.com/PostHog/posthog/issues/4705#issuecomment-895611473
- [x] Add current/previous tags in insight table
- [x] Make previous colors lighter corresponding current colors
- [x] Don't expose compare to if date filter is "All time"
- [x] Update date filter to show formatted dates (only changed for Insight Date Filter to reduce blast radius)
- [x] Add legend to trends + stickiness insights (insights with compare to functionality)
- [x] Fix various compare bugs
- [x] Add `compare_label` parameter to trend result
- [x] Add time freezing dev package
- [x] Show new legend by default with `legend_hidden` filter param
- [x] Fix breakdown visibilities in entity labels in various places: insights table, funnels step table, filters, etc.

Note: I'll tackle tooltips in another PR.

## Before

screencapture-app-posthog-insights-FFuRBl7B-edit-2021-12-06-16_14_04

## After - Can you spot the differences?

With compare to previous

screencapture-localhost-8000-insights-aJp403j8-edit-2021-12-08-11_29_13

With breakdown

screencapture-localhost-8000-insights-aJp403j8-edit-2021-12-08-11_29_56

Insight types this affects:
- [x] All stickiness graph types
- [x] All trends graph types
- [x] Ignore sessions (on deprecation path)

## How did you test this code?

Added logic tests + smoke testing of new ui components.

Works with breakdowns, multi breakdowns, $current_urls, etc.

There is a bug with today's code where breakdowns + compare doesn't return data from the previous time range (see bug in action [here](https://app.posthog.com/insights/FFuRBl7B/edit?insight=TRENDS&actions=%5B%5D&events=%5B%7B%22id%22%3A%22%24pageview%22%2C%22name%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A0%7D%2C%7B%22id%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A1%2C%22name%22%3A%22%24pageview%22%7D%2C%7B%22id%22%3A%22%24pageview%22%2C%22type%22%3A%22events%22%2C%22order%22%3A2%2C%22name%22%3A%22%24pageview%22%7D%5D&display=ActionsLineGraph&interval=day&properties=%5B%5D&funnel_to_step=1&funnel_viz_type=steps&exclusions=%5B%5D&funnel_from_step=0&compare=true&breakdown=%24current_url&breakdown_type=event)). This PR doesn't fix this, but theoretically should work once the backend bug is fixed (tested with dummy data). @EDsCODE  do you know if this is being tracked anywhere?

### CODE REVIEW

**Review by paolodamico — DISMISSED** [review]
Initial pass (UI/UX only, haven't looked much at the code), it's looking pretty great!

1. I'm not sure if the new banner fits with the current design. I would suggest either adding a light gray background or if we're going to add a background to the main graph (per the Figma design). Will defer to @clarkus here.
    
2. I don't think this is good UX, because it's hard to parse. One approach that comes to mind is changing the header name to show both relevant dates. An alternative is just say "Day n" in the header and show the date in the cell.
    
3. Unsure if it's part of the scope of this PR, but tooltips are still super confusing.
    
4. I can still select the "compare" option even if all time is set (which btw made my browser crash, so good call here).
    
5. Re date formats, not sure if we should zero-pad numbers, feels pretty common outside the US. Thoughts @clarkus ?

The new legend is amazing! Love it, some thoughts.
6. I think it might be worth vertically centering the legend card if it's small.
    
7. The behavior of opening the legend feels strange. Not sure if it's a matter of adding an animation. Any ideas @clarkus ? Also, thinking it might be worth having it open by default (as an experiment).
8. I found this bug when I created the chart in the screenshot where the legend shows empty.
    
9. Think the legend should have a max width and then just allow scrolling.
    image

**Comment by alexkim205** [comment]
Appreciate the quick review @paolodamico!

> 1. I'm not sure if the new banner fits with the current design. I would suggest either adding a light gray background or if we're going to add a background to the main graph (per the Figma design). Will defer to @clarkus here.

I've added a light grey bg to the insight graph container and standardized the bg colors we use across insights to bring it up to speed with our Figma mockups. 

> 2. I don't think this is good UX, because it's hard to parse. One approach that comes to mind is changing the header name to show both relevant dates. An alternative is just say "Day n" in the header and show the date in the cell.

I've changed it so that dates are now inside of each cell when compare functionality is on. Also fixed an existing bug where compare labels weren't being shown anywhere in our UX, which was making it difficult to distinguish current and previous rows. This let's me get rid of the compare column altogether. See screenshots above

> 3. Unsure if it's part of the scope of this PR, but tooltips are still super confusing.

Yup, i'd prefer to chunk this out into it's own PR to avoid making this bigger. I'll tag you when that's ready!

> 4. I can still select the "compare" option even if all time is set (which btw made my browser crash, so good call here).

Fixed and wrote tests for it

> 5. Re date formats, not sure if we should zero-pad numbers, feels pretty common outside the US. Thoughts @clarkus ?

my 2c, zero-padding here will make the width ever so slightly wider than it has to be. We're already working with limited real estate in the insight graph header which is something to keep in mind too if we plan to add more stuff there in the future.

> 6. I think it might be worth vertically centering the legend card if it's small. 

Done!

> 8. I found this bug when I created the chart in the screenshot where the legend shows empty. 

Fixed

 9. Think the legend should have a max width and then just allow scrolling. 

Max-height with overflow scrolling implemented.

**Comment by clarkus** [comment]
> I've added a light grey bg to the insight graph container and standardized the bg colors we use across insights to bring it up to speed with our Figma mockups.

👍 

> I've changed it so that dates are now inside of each cell when compare functionality is on. Also fixed an existing bug where compare labels weren't being shown anywhere in our UX, which was making it difficult to distinguish current and previous rows. This let's me get rid of the compare column altogether. See screenshots above

While I see the value of days of the week, I don't think we need to show them directly at all. We use tooltips to translate other dates in the app, so we could just layer in the day of the week into that translation. 

> my 2c, zero-padding here will make the width ever so slightly wider than it has to be. We're already working with limited real estate in the insight graph header which is something to keep in mind too if we plan to add more stuff there in the future.

:+1:

File: frontend/src/lib/components/icons.tsx [hunk 1]
```diff
@@ -916,3 +916,14 @@ export function IconKeyboard(props: React.SVGProps<SVGSVGElement>): JSX.Element
         </svg>
     )
 }
+
+export function LegendIcon({ style }: { style?: CSSProperties }): JSX.Element {
```
Comment (paolodamico) [inline frontend/src/lib/components/icons.tsx:920]: Nit: `IconLegend` (for better consistency)

File: frontend/src/scenes/insights/InsightContainer.tsx [hunk 2]
```diff
@@ -175,17 +176,40 @@ export function InsightContainer(): JSX.Element {
...
                     >
-                        <Col>
-                            <FunnelCanvasLabel />
-                            <PathCanvasLabel />
-                        </Col>
-                        {lastRefresh && <ComputationTimeWithRefresh />}
+                        {activeView === InsightType.FUNNELS ||
+                            (activeView === InsightType.PATHS && (
+                                <Col>
+                                    <FunnelCanvasLabel />
```
Comment (paolodamico) [inline frontend/src/scenes/insights/InsightContainer.tsx:182]: We have a double check for `activeView` in the component, we should avoid this for easier maintenance

File: frontend/src/lib/components/InsightLegend/InsightLegend.tsx [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import { formatCompareLabel } from 'scenes/insights/InsightsTable/InsightsTable'
+
+export function InsightLegendButton(): JSX.Element {
+    const { filters } = useValues(insightLogic)
+    const { toggleInsightLegend } = useActions(insightLogic)
+
+    return (
+        <Button className="insight-legend-button" onClick={toggleInsightLegend}>
+            <LegendIcon />
+            <span className="insight-legend-button-title">{filters.legend_hidden ? 'Show' : 'Hide'} Legend</span>
```
Comment (paolodamico) [inline frontend/src/lib/components/InsightLegend/InsightLegend.tsx:21]:
```suggestion
            <span className="insight-legend-button-title">{filters.legend_hidden ? 'Show' : 'Hide'} legend</span>
```

**Review by paolodamico — APPROVED** [review]
It's a bit hard to fully evaluate such a large PR and make sure it doesn't introduce any new bugs, in general looks good to me! Some improvement suggestions below, but I'd suggest considering a separate PR,

- I adjusted the UI of the legend table with compare a bit, but I think we need @clarkus's eye here.
- Ideally we would sort the legend table with "Current" & "Previous" for each event/action in the same order (see different order here) for easy comparison.

File: frontend/src/lib/components/DateFilter/DateFilter.tsx [hunk 4]
```diff
@@ -127,19 +136,30 @@ export function DateFilter({
...
                     }
 
-                    if (inactive && parsedValue !== key) {
+                    if (inactive && currKey !== key) {
                         return null
                     }
 
                     return (
-                        <Select.Option key={key} value={key} label={makeLabel ? makeLabel(key) : undefined}>
+                        <Select.Option
```
Comment (rcmarron) [inline frontend/src/lib/components/DateFilter/DateFilter.tsx:152]: I don't know the context for this change, but it feels weird to select an item in a drop down and then not have the drop down show the thing I clicked it. If we want to show the full date string, I wonder if it should be shown along with the range that the user clicked?

File: frontend/src/lib/components/InsightLegend/InsightLegend.tsx [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        <div className={clsx('insight-legend-menu', indexedResults.length <= 5 && 'short')}>
+            {indexedResults &&
+                indexedResults.map((item) => {
+                    return (
+                        <Row key={item.id} className="insight-legend-menu-item" wrap={false}>
+                            <Col>
+                                <PHCheckbox
+                                    color={colorList[item.id]}
+                                    checked={visibilityMap[item.id]}
+                                    showIcon={indexedResults.length > 1}
```
Comment (rcmarron) [inline frontend/src/lib/components/InsightLegend/InsightLegend.tsx:43]: Feels kind of like we're abusing the PHCheckbox by disbaling it and removing the checkbox. Should we just draw a box instead of adding the `showIcon` prop?

File: frontend/src/scenes/insights/InsightTabs/InsightDisplayConfig.tsx [hunk 6]
```diff
@@ -92,60 +92,76 @@ export function InsightDisplayConfig({
...
+                            defaultValue={getFormattedLastWeekDate()}
                             disabled={dateFilterDisabled}
-                            bordered={false}
+                            bordered
                             makeLabel={(key) => (
                                 <>
                                     <CalendarOutlined /> {key}
                                 </>
                             )}
+                            isDateFormatted
```
Comment (rcmarron) [inline frontend/src/scenes/insights/InsightTabs/InsightDisplayConfig.tsx:108]: Same thought on the date selector being odd here.

File: frontend/src/lib/components/InsightLegend/InsightLegend.tsx (hunk 5 above)
Comment (rcmarron) [reply frontend/src/lib/components/InsightLegend/InsightLegend.tsx:43]: On second thought - it doesn't feel too weird to just leave the checkbox functionality even if there is only one item. Maybe that's the right thing here?

File: frontend/src/lib/components/InsightLegend/InsightLegend.tsx [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    const { insightProps, filters } = useValues(insightLogic)
+    const logic = trendsLogic(insightProps)
+    const { indexedResults, visibilityMap } = useValues(logic)
+    const { toggleVisibility } = useActions(logic)
+    const colorList = getChartColors('white', indexedResults.length, !!filters.compare)
+
+    return (
+        <div className={clsx('insight-legend-menu', indexedResults.length <= 5 && 'short')}>
+            {indexedResults &&
+                indexedResults.map((item) => {
```
Comment (rcmarron) [inline frontend/src/lib/components/InsightLegend/InsightLegend.tsx:36]: I'm not sure why this is happening, but the legend has a different order than the table below. You can see this if you turn on compare previous. The order here seems better, but either way they should probably be consistent.

**Review by rcmarron — DISMISSED** [review]
A lot of good fixes in here! A few thoughts in the comments.

File: frontend/src/lib/components/InsightLegend/InsightLegend.tsx (hunk 5 above)
Comment (alexkim205) [reply frontend/src/lib/components/InsightLegend/InsightLegend.tsx:43]: The reason I changed it to hide the check was because I found myself clicking the checkbox trying to toggle it on and off. Maybe it makes sense not to disabled this on single series data altogether? It doesn't make sense that the user shouldn't have to power to hide everything if they wanted to?

File: frontend/src/lib/components/CompareFilter/compareFilterLogic.ts [hunk 8]
```diff
@@ -26,6 +27,16 @@ export const compareFilterLogic = kea<compareFilterLogicType>({
         ],
     }),
     listeners: ({ actions, values }) => ({
+        init: ({ searchParams: { compare, date_from, insight } }) => {
+            if (compare !== undefined) {
+                actions.setCompare(compare)
+            }
+            if (insight === InsightType.LIFECYCLE || date_from === 'all') {
```
[the remark this answers is not in the document]
Comment (alexkim205) [reply frontend/src/lib/components/CompareFilter/compareFilterLogic.ts:34]: good call!

File: frontend/src/lib/components/InsightLegend/InsightLegend.tsx (hunk 5 above)
Comment (rcmarron) [reply frontend/src/lib/components/InsightLegend/InsightLegend.tsx:43]: Yeah, not disabling on a single series makes sense to me.

File: frontend/src/lib/components/InsightLegend/InsightLegend.tsx (hunk 7 above)
Comment (alexkim205) [reply frontend/src/lib/components/InsightLegend/InsightLegend.tsx:36]: I think I figured it out. The table comes presorted by total sum column and canceling this returns the table to it's original correct order. I've changed it so that we don't sort by total sum by default.

**Review by clarkus — DISMISSED** [review]
How is the size of the legend being determined? There's a good bit of extra space in this example and I don't see any content that would justify the space - just curious if we can optimize this more. Likewise I'm not sure if needs to be full height. I think it hsould size ot its contents. If those contents are taller than the available area, we can scroll the legend vertically. 

Screen Shot 2021-12-08 at 11 37 49 AM

This date formatting really competes with reading the metric values. I don't think we need to show the day of the week prominently here. Instead we should focus on showing the current time frame prominently with the comparison range in secondary position. For example a general pattern could be `Current (Comparison)` or **Dec 1** (Nov 23).

Screen Shot 2021-12-08 at 11 40 51 AM

I think we should swap the legend values when breakdowns are applied. The table below shows each browser value primarily. The legend should match this. The most meaningful identifier is the value, not the event. 

Screen Shot 2021-12-08 at 11 47 06 AM

**Comment by EDsCODE** [comment]
RE: bug with breakdowns + compare. The link isn't showing me a graph that's broken or has compare set. I also just tried a breakdown with compare active and it seems to return as expected. Let me know what I should look at for the bug

**Comment by alexkim205** [comment]
> RE: bug with breakdowns + compare. The link isn't showing me a graph that's broken or has compare set. I also just tried a breakdown with compare active and it seems to return as expected. Let me know what I should look at for the bug

Huh it seems to be working now, looks like it was transient. Thanks for taking a look anyways!

**Comment by alexkim205** [comment]
Thanks @clarkus! Implemented all your feedback and this is what it looks like now:


Screen Shot 2021-12-08 at 4 42 59 PM

Screen Shot 2021-12-08 at 4 42 21 PM

Screen Shot 2021-12-08 at 4 42 14 PM

Screen Shot 2021-12-08 at 4 42 03 PM
